pkg/compact/compact.go

```diff
@@ -0,0 +1,544 @@
+			level.Warn(c.logger).Log("msg", "delete outdated block from group", "block", id, "err", err)
+		}
+	}
+	return nil
+}
+
+func (c *Syncer) groupKey(meta *block.Meta) string {
+	return fmt.Sprintf("%x", labels.FromMap(meta.Thanos.Labels).Hash())
+}
+
+// add adds the block in the given directory to its respective compaction group.
+func (c *Syncer) add(bdir string) error {
+	meta, err := block.ReadMetaFile(bdir)
+	if err != nil {
+		return errors.Wrap(err, "read meta file")
+	}
+	h := c.groupKey(meta)
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	g, ok := c.groups[h]
+	if !ok {
+		g, err = NewGroup(
+			log.With(c.logger, "compactionGroup", h),
+			c.bkt,
+			filepath.Join(c.dir, "groups", h),
+			meta.Thanos.Labels,
+		)
+		if err != nil {
+			return errors.Wrap(err, "create compaction group")
+		}
+		c.groups[h] = g
+	}
+	return g.Add(meta)
+}
+
+// Groups returns the compaction groups created by the Syncer.
+func (c *Syncer) Groups() (res []*Group) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	for _, g := range c.groups {
+		res = append(res, g)
+	}
+	return res
+}
+
+// GarbageCollect deletes blocks from the bucket if their data is available as part of a
+// block with a higher compaction level.
+func (c *Syncer) GarbageCollect(ctx context.Context) error {
+	// Map each block to its highest priority parent. Initial blocks have themselves
+	// in their source section, i.e. are their own parent.
+	var (
+		all     = map[ulid.ULID]*block.Meta{}
+		parents = map[ulid.ULID]ulid.ULID{}
+	)
+	err := iterBlocks(c.dir, func(dir string, _ ulid.ULID) error {
+		meta, err := block.ReadMetaFile(dir)
+		if err != nil {
+			return errors.Wrap(err, "read meta")
+		}
+		all[meta.ULID] = meta
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	for id, meta := range all {
+		// For each source block we contain, check whether we are the highest priotiy parent block.
+		for _, sid := range meta.Compaction.Sources {
+			pid, ok := parents[sid]
+			// No parents for the source block so far.
+			if !ok {
+				parents[sid] = id
+				continue
+			}
+			pmeta, ok := all[pid]
+			if !ok {
+				return errors.Errorf("previous parent block %s not found", pid)
+			}
+			// The current block is the higher priority parent for the source if its
+			// compaction level is higher than that of the previously set parent.
+			// If compaction levels are equal, the more recent ULID wins.
+			//
+			// The ULID recency alone is not sufficient since races, e.g. induced
+			// by downtime of garbage collection, may re-compact blocks that are
+			// were already compacted into higher-level blocks multiple times.
+			level, plevel := meta.Compaction.Level, pmeta.Compaction.Level
+
+			if level > plevel || (level == plevel && id.Compare(pid) > 0) {
+				parents[sid] = id
+			}
+		}
+	}
+
+	// A block can safely be deleted if they are not the highest priority parent for
+	// any source block.
+	hasChildren := map[ulid.ULID]struct{}{}
```

Comment:
ok, can we then rename this `hasChildren` to just `topParents`? That would help to understand what's going on -> all intermediate blocks have children as well (: This what was most confusing here